From 577fdcde831cd82ceafa325edacbd1b19146c45d Mon Sep 17 00:00:00 2001
From: Tommy Haung <tommy_huang@aspeedtech.com>
Date: Mon, 27 Dec 2021 08:57:39 +0800
Subject: [PATCH 18/18] drivers:i2c:Add i2c M/S controller issue

When the i2c master and slave are all tuned on at same time.
The i2c master package done will be observed by slave and lock.
Reset i2c salve to patch this condition.

Signed-off-by: Tommy Haung <tommy_huang@aspeedtech.com>
Change-Id: Idf441e8e4cdace8b546a69a34a1c8177508be2f3
---
 drivers/i2c/i2c_aspeed.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/i2c_aspeed.c b/drivers/i2c/i2c_aspeed.c
index 8192e974d4..f289bc81ff 100644
--- a/drivers/i2c/i2c_aspeed.c
+++ b/drivers/i2c/i2c_aspeed.c
@@ -1210,7 +1210,18 @@ int aspeed_i2c_master_irq(const struct device *dev)
 			k_sem_give(&data->sync_sem);
 			break;
 		case AST_I2CM_TX_ACK:
-			/*LOG_DBG("M : AST_I2CM_TX_ACK = %x\n", sts);*/
+#ifdef CONFIG_I2C_SLAVE
+			/* Workaround for master/slave package mode enable rx done stuck issue
+			 * When master go for first read (RX_DONE), slave mode will also effect
+			 * Then controller will send nack, not operate anymore.
+			 */
+			if (sys_read32(i2c_base + AST_I2CS_CMD_STS) & AST_I2CS_PKT_MODE_EN) {
+				uint32_t slave_cmd = sys_read32(i2c_base + AST_I2CS_CMD_STS);
+
+				sys_write32(0, i2c_base + AST_I2CS_CMD_STS);
+				sys_write32(slave_cmd, i2c_base + AST_I2CS_CMD_STS);
+			}
+#endif
 		case AST_I2CM_TX_ACK | AST_I2CM_NORMAL_STOP:
 			LOG_DBG("M : I2CM_TX_ACK | I2CM_N_S = %x\n", sts);
 			do_i2cm_tx(dev);
-- 
2.17.1

